2025-07-11-general-sso.md

```diff
@@ -0,0 +1,210 @@
+```yaml
+STORJ_SSO_OIDC_PROVIDER_INFOS: general-google:client-id,client-secret,https://accounts.google.com;storj-okta:client-id,client-secret,https://okta-domain.com
+```
+We do not need to change the `EmailProviderMappings` configuration, as general providers will not use email matching.
+
+**NB**: having no provider with the "general-" prefix effectively disables general SSO functionality.
+
+#### What to Test
+- **Google OAuth2 flow**: we have to make sure the Google SSO flow works correctly;
+  - log in to an existing account (account linking)
+  - log in to a new account
+- **Enterprise SSO flow**: test that all existing enterprise SSO flows continue to work as expected.
+- **Enterprise SSO precedence**:
+  - a user with an email matching an enterprise SSO provider should always be redirected to that provider, even if general SSO is available.
+  - a user already linked to an enterprise SSO provider should not be able to log in with general SSO or password or vice versa.
+  - a user with an email matching an enterprise SSO provider should not be able to log in with general SSO or password.
+- Password reset/change functionality
+- Account deletion functionality
+- Account linking
+
+## Implementation Tickets
+
+1. **[SATELLITE] Extend SSO Service for General Providers**
+  - Modify the code in `satellite/console/consoleauth/sso/service.go` to support general SSO providers
+  - General providers are identified by the "general-" prefix.
+  - Update the /api/v0/auth/sso/url endpoint to return general SSO options when no enterprise provider matches the email.
+  - Update restrictions to allow general SSO users to access functionality current SSO are not allowed to access,
+    such as change/reset password and delete account flows, except changing email.
+
+2. **[SATELLITE UI] Update Login Flow UI**
+  - For each general SSO provider, received from the /sso/url endpoint,
+    add a button that redirects to that provider's authentication flow.
+
+3. **[SATELLITE UI] Allow general SSO users to manage their account fully**
+  - If a user is linked to a general SSO provider,
+    - Show a set password option in account settings if they have no password
+    - Show the existing change password option otherwise.
+    - Allow them to delete their account.
+
+4. **Set up Google SSO**
+  - Set up an external Google OAuth2 application in GCP
+  - Add the client ID and secret to the `STORJ_SSO_OIDC_PROVIDER_INFOS` configuration
+
+## Open Questions
+* Should we allow users to unlink their SSO accounts and revert to email/password only?
+* Should we allow general SSO users to change their email?
+* Should we store state tokens in the db instead of cookies?
+* Should we require extra validation for linking a general SSO account to an existing user?
+  * What happens if a user's SSO provider account is compromised? Account linking should be secure and reversible
+* Should link enterprise SSO accounts to users who have already authenticated with general SSO?
```

Comment:
if the enterprise SSO email mappings match an email, it should override/replace general SSO